File: hw02/Dog.java
package hw02;

import java.util.Date;
import java.util.List;

/**
 * Cat is a class inherited from classes Animal -> Pet
 */
public class Dog extends Pet{
    protected boolean trained;

    /**
     * It is a constructor for Dog class
     * @param height            The height of animal
     * @param weight            The weight of animal
     * @param eyeColor          The eye color of animal
     * @param name              The name of animal
     * @param breed             The breed of animal
     * @param vaccinations      The vaccinations list of animal
     * @param furColor          The fur color of animal
     * @param birthDate         The birthdate of animal
     * @param trained           The trained status of animal
     */
    public Dog(double height, double weight, String eyeColor, String name, String breed,
            List<String> vaccinations, String furColor, Date birdhDate, boolean trained) {
        super(height, weight, eyeColor, name, breed, vaccinations, furColor, birdhDate);
        this.trained = trained;
    }
    
    /**
     * It is a constructor for Dog class (trained status will be false)
     * @param height            The height of animal
     * @param weight            The weight of animal
     * @param eyeColor          The eye color of animal
     * @param name              The name of animal
     * @param breed             The breed of animal
     * @param vaccinations      The vaccinations list of animal
     * @param furColor          The fur color of animal
     * @param birthDate         The birthdate of animal
     * @param trained           The trained status of animal
     */
    public Dog(double height, double weight, String eyeColor, String name, String breed,
            List<String> vaccinations, String furColor, Date birdhDate) {
        super(height, weight, eyeColor, name, breed, vaccinations, furColor, birdhDate);
        this.trained = false;
    }

    /**
     * Returns is the animal trained or not
     * @return this.trained
     */
    public boolean isTrained() {
        return this.trained;
    }
    
    public void train() {
        this.trained = true;
        System.out.printf("The dog %s was trained successfully\n", this.getName());
    }

    @Override
    public void sound() {
        System.out.println("Woof-woof!");
    }
    
    @Override
    public String toString() {
        return String.format("Dog (%s, is trained: %b)", super.toString(), this.isTrained());
    }
}

File: hw07/internal/models/dog.go
package models

import (
	"fmt"
	"time"
)

// Dog structure
type Dog struct {
	Pet
	trained bool
}

// Creates new Dog instance with full parameters given
func NewDog(height float64, weight float64, eyeColor string, name string, breed string, vaccinations []string, furColor string, birthDate time.Time, trained bool) (*Dog, error) {
	p, err := newPet(height, weight, eyeColor, name, breed, vaccinations, furColor, birthDate)
	if err != nil {
		return nil, err
	}

	c := &Dog{
		*p,
		trained,
	}

	// Handles potential restrictions of any parameter to return error when new Dog cannot be instantiated
	return c, nil
}

// Returns the is dog trained
func (d *Dog) IsTrained() bool {
	return d.trained
}

// Trains the dog
func (d *Dog) Train() string {
	d.trained = true
	return fmt.Sprintf("%s is trained", d.name)
}

// Returns sound of dog
func (d *Dog) Sound() string {
	return fmt.Sprint("Woof-woof!")
}

// Returns the string representation of Dog instance
func (d *Dog) String() string {
	return fmt.Sprintf("Dog %v, is trained: %t", d.Pet.String(), d.trained)		
}